File: memory-domain/src/main/java/com/memory/domain/comment/repository/CommentRepositoryCustom.java
package com.memory.domain.comment.repository;

import com.memory.domain.comment.Comment;
import com.memory.domain.member.Member;
import com.memory.domain.memory.Memory;

import java.util.List;
import java.util.Optional;

public interface CommentRepositoryCustom {

    List<Comment> findCommentsByMemoryWithHierarchy(Memory memory);

    List<Comment> findTopLevelCommentsByMemory(Memory memory, int page, int size);

    List<Comment> findRepliesByParentComment(Comment parentComment);

    long countActiveCommentsByMemory(Memory memory);

    List<Comment> findCommentsByMember(Member member, int page, int size);

    Optional<Comment> findByIdWithMember(Long commentId);

    List<Comment> findRecentComments(int limit);

    Optional<Comment> findCommentById(Long commentId);
}

File: memory-domain/src/main/kotlin/com/memory/domain/comment/repository/CommentRepositoryCustom.kt
package com.memory.domain.comment.repository

import com.memory.domain.comment.Comment
import com.memory.domain.member.Member
import com.memory.domain.memory.Memory
import java.util.Optional

interface CommentRepositoryCustom {

    fun findCommentsByMemoryWithHierarchy(memory: Memory?): List<Comment>

    fun findTopLevelCommentsByMemory(memory: Memory?, page: Int, size: Int): List<Comment>

    fun findRepliesByParentComment(parentComment: Comment?): List<Comment>

    fun countActiveCommentsByMemory(memory: Memory?): Long

    fun findCommentsByMember(member: Member?, page: Int, size: Int): List<Comment>

    fun findByIdWithMember(commentId: Long?): Optional<Comment>

    fun findRecentComments(limit: Int): List<Comment>

    fun findCommentById(commentId: Long?): Optional<Comment>
}
